File: infra/supabase_client.py
"""
Supabase client configuration and utilities.

This module provides a singleton Supabase client instance
and utility functions for interacting with Supabase services.
"""
import os
from supabase import create_client, Client
from typing import Optional


_supabase_client: Optional[Client] = None


def get_supabase_client() -> Client:
    """
    Get or create a Supabase client instance.
    
    This function implements a singleton pattern to reuse
    the same client instance across the application.
    
    Returns:
        Client: Configured Supabase client
        
    Raises:
        ValueError: If SUPABASE_URL or SUPABASE_KEY are not set
    """
    global _supabase_client
    
    if _supabase_client is None:
        url = os.getenv("SUPABASE_URL")
        key = os.getenv("SUPABASE_KEY")
        
        if not url or not key:
            raise ValueError(
                "SUPABASE_URL and SUPABASE_KEY must be set in environment variables"
            )
        
        _supabase_client = create_client(url, key)
    
    return _supabase_client

File: src/frontend/src/services/supabase.client.ts
/**
 * Supabase Client Singleton (T-031-FRONT)
 * 
 * Provides a singleton instance of the Supabase client for Realtime subscriptions.
 * Supports Dependency Injection for testing and custom configurations.
 */

import { createClient, SupabaseClient } from '@supabase/supabase-js';

/**
 * Configuration for Supabase client initialization.
 */
export interface SupabaseConfig {
  url: string;
  anonKey: string;
}

let supabaseInstance: SupabaseClient | null = null;

/**
 * Get or create the Supabase client instance (singleton pattern).
 * 
 * Supports two modes:
 * 1. Production: Reads from import.meta.env (default)
 * 2. Testing/Custom: Accepts explicit config parameter
 * 
 * @param config - Optional configuration object for DI/testing
 * @throws Error if configuration is missing or invalid
 * @returns Supabase client instance
 */
export function getSupabaseClient(config?: SupabaseConfig): SupabaseClient {
  // Return existing instance if already created
  if (supabaseInstance) {
    return supabaseInstance;
  }

  // Determine configuration source
  const finalConfig: SupabaseConfig = config || {
    url: import.meta.env.VITE_SUPABASE_URL,
    anonKey: import.meta.env.VITE_SUPABASE_ANON_KEY,
  };

  // Validate configuration
  if (!finalConfig.url || !finalConfig.anonKey) {
    throw new Error(
      'Missing Supabase environment variables. Please configure VITE_SUPABASE_URL and VITE_SUPABASE_ANON_KEY.'
    );
  }

  // Create and store singleton instance
  supabaseInstance = createClient(finalConfig.url, finalConfig.anonKey);

  return supabaseInstance;
}

/**
 * Reset the singleton instance (for testing purposes only).
 * @internal
 */
export function resetSupabaseClient(): void {
  supabaseInstance = null;
}
